### ISSUE TITLE
Support for Bucket.location setter

### ISSUE THREAD

**Issue by gastlich** [GoogleCloudPlatform/google-cloud-python#5194]
I was trying to change location of the bucket in Storage, but everything gets updated apart from location. It looks like the data is being sent through API, but are not persisted?

Code example
```python
storage_client = StorageClient()
bucket = storage_client.create_bucket(name)
bucket.storage_class = 'MULTI_REGIONAL'
bucket.location = 'EU' # tried 'eu' as well
bucket.labels = {
     'project': 'etl-transformer',
     'group': 'etl',
}
bucket.patch()
```

```python
# update_properties printed in _PropertyMixin:patch
{
    'labels': {
        'project': 'etl-transformer',
        'group': 'etl'
    },
    'storageClass': 'MULTI_REGIONAL',
    'location': 'EU'
}

# I checked here even dump in requests package to be sure it's being passed to the http object

# api_reponse
{
    u 'labels': {
      u 'project': u 'etl-transformer',
      u 'group': u 'etl'
    },
    u 'location': u 'US',
    u 'storageClass': u 'MULTI_REGIONAL'
}
```

I know that documentation doesn't say that `location` attribute is a setter, but under the hood it's a `_scalar_property` which is appended to changed properties and according to json API, location is writable:

https://cloud.google.com/storage/docs/json_api/v1/buckets

**Comment by frankyn** [comment]
Hi @gastlich,

Could you try the following instead:
```python
# Imports the Google Cloud client library
from google.cloud import storage

# Instantiates a client
storage_client = storage.Client()

# The name for the new bucket
bucket_name = 'bucket-name'

# Create a new bucket object
bucket_object = storage.bucket.Bucket(storage_client, bucket_name)

# Set location
bucket_object.location = "EU"

# Set storage class
bucket_object.storage_class = "MULTI_REGIONAL"

# Create bucket
bucket_object.create()
```

**Comment by gastlich** [comment]
Hi @frankyn ,
it worked, I thought there is only `create_bucket` way to create buckets, didn't realise, that there might be just `.bucket()` constructor...

but still, do you know the reason, why it hasn't been updated in my first example?

Cheers

**Comment by crwilcox** [comment]
@gastlich,

The issue is each call does something a bit different. storage_client.bucket creates a bucket object on the client. storage_client.create_bucket creates the bucket on the platform. When you changed your code to use bucket() you changed it to not actually create anything until you had set the location to EU.

In the original you had created a bucket in US and then tried to change its location to EU. Once created, you cannot change the location of a bucket.

**Comment by tseaver** [comment]
Two code changes are required to address this issue:

- Plumb the `location` argument through to `Client.create_bucket`.
- Make the `Bucket.location` property read-only.

**Comment by tseaver** [comment]
Hmmm, making `Bucket.location` read-only in all cases would be backward-incompatible:  applications which set the location between instantiating the bucket and calling its `create` method would break.

**Comment by crwilcox** [comment]
That could be done, but the comment for `create_bucket` already states that for more complex creation you should use the other flow. My guess would be this is because if we add location, really we need to start adding all of the possible configuration options to `create_bucket` to support all uses. But we already have an alternative for that.

The doc comment I am referring to btw:

        To set additional properties when creating a bucket such as
        :attr:`~.Bucket.location`, use :meth:`~.Bucket.create`.

**Comment by tseaver** [comment]
@crwilcox OK.  Maybe the setter for the `location` property should still raise an error if there is already a value set?

**Comment by crwilcox** [comment]
That seems like a reasonable idea. Really that entire class is *almost* entirely read-only once written fields from what I could tell. That fix would help to prevent this in the future.

**Comment by tseaver** [comment]
> Really that entire class is almost entirely read-only once written fields from what I could tell.

According to the API docs for [`buckets.patch`](https://cloud.google.com/storage/docs/json_api/v1/buckets/patch), there are a bunch of writable bucket properties.

**Comment by crwilcox** [comment]
That may have been a strong statement. The more accurate statement I should have made is that in fixing this we should be sure we do the same protection of all write-once settings.

**Comment by tseaver** [comment]
`Bucket.location` is the only property set using the `_scalar_property` helper, which gives it a non-validating setter.

The following properties are already read-only (no setter defined):

- `path`
- `client`
- `user_project`
- `acl` (but the `ACL` itself is mutable).
- `default_object_acl` (likewise)
- `etag`
- `id`
- `metageneration`
- `owner`
- `project_number`
- `self_link`
- `time_created`

The following bucket properties have explicit setters:

- `cors`
- `labels`
- `lifecycle_rules`
- `storage_class`
- `versioning_enabled`
- `requester_pays`

The `logging` and `website` resource keys are supported through non-property mutators.

**Comment by tseaver** [comment]
Closed via #5808.
